accepted/2023/simplify-output-paths.md

```diff
@@ -0,0 +1,140 @@
+# Simplify .NET Output Paths
+
+We'd like to make some improvements to project output paths for .NET 8.
+
+Goals:
+
+- Support output paths in a repo or solution-specific folder, rather than each project folder having its own output paths
+- Provide simple and consistent output paths
+- Avoid cases where one output path is nested inside of another one
+- Avoid excessively deep directory heirarchies
+- Allow output path customization when needed
+
+## Motivation
+
+Some developers would like to put all the output for their solution or repo [under a single folder](https://github.com/dotnet/sdk/issues/867).  This would make it easier to delete all of the output, as MSBuild-based clean only deletes output for a single configuration, and doesn't delete everything from the intermediate output folder, as that's where it keeps track of what was produced by the build and hence what should be deleted in a clean.  It would support building with [read only](https://github.com/dotnet/sdk/issues/887) [source trees](https://github.com/dotnet/designs/pull/281#discussion_r1072949097).  It could also make it easier to set up CI and deployment.
+
+The current output paths for .NET projects are rather complicated and inconsistent.  The default output path for a .NET project includes folders for the `TargetFramework` and the `RuntimeIdentifier` (if it is set).  Additionally, the publish output goes in a `publish` subfolder of the output folder.  The resulting paths are as follows:
+
+- `bin\<Configuration>\<TargetFramework>` - Build output with no `RuntimeIdentifier`
```

Comment:
Here's a crazy idea to think about, what about changing the build output directory structure to match that of nuget packages? `bin\lib\net6.0\mylib.dll`, `bin\lib\netstandard2.0\mylib.dll`, etc.  Maybe only for projects with `<Packable>true`?

Today, it's very difficult for package authors to test their packages that aren't simple class libraries. If their package contains `contentFiles/` or `build/` or `buildTransitive/`, they have no choice but to pack their nupkg, then use a local package source to install the package via NuGet in a test project. And then clean the global package directory every time they want to make a change in the package.

Even if we can solve how to import a project's msbuild props/targets file via ProjectReference, the different directory structure of project bin\ vs package structure would make it difficult for package authors to write props/targets that work in both PackageReference and ProjectReference.

Hence, if we can more closely align the build output directory structure to NuGet packages, maybe it will make debugging NuGet packages via ProjectReference more feasible.

Reply:
This way, you could technically merge `publish` and `bin` into a single final target instead of having an additional build step to copy `bin` to `publish`. That leaves with the `pack` folder. So, essentially we'll end up with the 4 folders instead of 5 and we can keep everything under single folder.

- **Extensions** folder (_where NuGet restores and other project extensions go_)
- **Interim** folder (_where intermediate and temporary files go_)
- **Final** folder (_where the published output goes_)
- **Publish/Package** folder (_where packages go_)

You could also have a single or multiple target folder across projects and their build contexts. That is MY light at the end of this tunnel! 💯